presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/ShardOrganizerUtil.java

```diff
@@ -13,49 +13,186 @@
  */
 package com.facebook.presto.raptor.storage.organization;
 
+import com.facebook.presto.raptor.metadata.MetadataDao;
 import com.facebook.presto.raptor.metadata.ShardMetadata;
+import com.facebook.presto.raptor.metadata.Table;
+import com.facebook.presto.raptor.metadata.TableColumn;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.type.TimestampType;
 import com.facebook.presto.spi.type.Type;
+import com.google.common.base.Joiner;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMultimap;
 import com.google.common.collect.Multimaps;
+import io.airlift.log.Logger;
+import org.skife.jdbi.v2.IDBI;
 
+import java.sql.Connection;
+import java.sql.JDBCType;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
 import java.time.Duration;
 import java.util.Collection;
-import java.util.Set;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.StringJoiner;
 
+import static com.facebook.presto.raptor.RaptorErrorCode.RAPTOR_ERROR;
+import static com.facebook.presto.raptor.metadata.DatabaseShardManager.maxColumn;
+import static com.facebook.presto.raptor.metadata.DatabaseShardManager.minColumn;
+import static com.facebook.presto.raptor.metadata.DatabaseShardManager.shardIndexTable;
+import static com.facebook.presto.raptor.metadata.JdbcUtil.jdbcType;
+import static com.facebook.presto.raptor.util.DatabaseUtil.metadataError;
 import static com.facebook.presto.spi.type.DateType.DATE;
+import static com.google.common.base.Verify.verify;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static com.google.common.collect.Iterables.partition;
+import static com.google.common.collect.Maps.uniqueIndex;
+import static io.airlift.slice.Slices.wrappedBuffer;
+import static java.util.Collections.nCopies;
+import static java.util.stream.Collectors.toList;
 
 public class ShardOrganizerUtil
 {
+    private static final Logger log = Logger.get(ShardOrganizerUtil.class);
+
     private ShardOrganizerUtil() {}
 
-    static Collection<Collection<ShardMetadata>> getShardsByDaysBuckets(Set<ShardMetadata> shardMetadata, Type type)
+    static Collection<ShardIndexInfo> toShardIndexInfo(
+            IDBI dbi,
+            MetadataDao metadataDao,
+            Table tableInfo,
+            Collection<ShardMetadata> shards,
+            boolean includeSortColumns)
     {
-        // bucket shards by the start day
-        ImmutableMultimap.Builder<Long, ShardMetadata> shardsByDays = ImmutableMultimap.builder();
+        Map<Long, ShardMetadata> shardsById = uniqueIndex(shards, ShardMetadata::getShardId);
+        long tableId = tableInfo.getTableId();
+
+        // include temporal range if present
+        String temporalColumnString = "";
```

Comment:
Building up the SELECT list with multiple strings is messy and error prone. Use list and join it at the end.